app/scripts/lib/setupWeb3.js

```diff
@@ -31,8 +36,117 @@ export default function setupWeb3(log) {
     value: web3.eth,
   })
 
+  // Setup logging of nested property usage
+  if (shouldLogUsage) {
+    // web3 namespaces with common and uncommon dapp actions
+    const includedTopKeys = [
+      'eth',
+      'db',
+      'shh',
+      'net',
+      'personal',
+      'bzz',
+      'version',
+    ]
+
+    // For each top-level property, create appropriate Proxy traps for all of
+    // their properties
+    includedTopKeys.forEach((topKey) => {
+      const applyTrapKeys = new Map()
+      const getTrapKeys = new Map()
+
+      Object.keys(web3[topKey]).forEach((key) => {
+        const path = `web3.${topKey}.${key}`
+
+        if (web3Entitites[path]) {
+          if (web3Entitites[path] === 'function') {
+            applyTrapKeys.set(key, path)
+          } else {
+            getTrapKeys.set(key, path)
+          }
+        }
+      })
+
+      // Create apply traps for function properties
+      for (const [key, path] of applyTrapKeys) {
+        web3[topKey][key] = new Proxy(web3[topKey][key], {
+          apply: (...params) => {
+            window.ethereum.request({
+              method: 'metamask_logInjectedWeb3Usage',
+              params: [
+                {
+                  action: 'apply',
+                  path,
+                },
+              ],
+            })
+
+            // Call function normally
+            return Reflect.apply(...params)
+          },
+        })
+      }
+
+      // Create get trap for non-function properties
+      web3[topKey] = new Proxy(web3[topKey], {
+        get: (web3Prop, key, ...params) => {
+          const name = stringifyKey(key)
+
+          if (getTrapKeys.has(name)) {
+            window.ethereum.request({
+              method: 'metamask_logInjectedWeb3Usage',
+              params: [
+                {
+                  action: 'get',
+                  path: getTrapKeys.get(name),
+                },
+              ],
+            })
+          }
+
+          // return value normally
+          return Reflect.get(web3Prop, key, ...params)
+        },
+      })
+    })
+  }
+
+  const unEnumerableFunctionKeys = [
```

Comment:
Why `unEnumerable`? :thinking: These properties do indeed seem to be enumerable.

Reply:
They don't show up on `Object.keys(web3)` on my end.

Reply:
Huh. Same here, but they _do_ show up if I run this:

```
for (const key of window.web3) {
  console.log(key)
}
```

:thinking: Doesn't "enumerable" mean "will be visited when iterating"? And this is iterating?

Reply:
>*Enumerable properties show up in for...in loops unless the property's key is a Symbol*. 

![image](https://user-images.githubusercontent.com/4448075/98296437-869e7400-1f78-11eb-94aa-093a5daf4862.png)

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties

Reply:
Ah I see - the difference is that these are _inherited_ enumerable properties, which show up in `for..in` but not in `Object.keys`

Reply:
Ah, nice. Thanks @brad-decker.

Where does that leave this PR?

Reply:
I mean, it's not of critical importance, but I'd suggest that a variable name of `topLevelFunctions` might be more clear in this case. Since these are in fact enumerable, and whether they are enumerable or not really isn't of particular importance here.